api/src/Validator/ContentNode/AssertPrototypeCompatibleValidator.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace App\Validator\ContentNode;
+
+use App\Entity\ContentNode;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+class AssertPrototypeCompatibleValidator extends ConstraintValidator {
+    public function validate($value, Constraint $constraint) {
+        if (!$constraint instanceof AssertPrototypeCompatible) {
+            throw new UnexpectedTypeException($constraint, AssertPrototypeCompatible::class);
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        $object = $this->context->getObject();
+        if (!$object instanceof ContentNode) {
+            throw new UnexpectedValueException($object, ContentNode::class);
+        }
+
+        if (!$value instanceof ContentNode) {
+            throw new UnexpectedValueException($value, ContentNode::class);
+        }
+
+        $class = get_class($object);
+
+        if (!is_a($value, $class, false)) {
+            $this->context->buildViolation($constraint->messageClassMismatch)
+                ->setParameter('{{ expectedClass }}', $class)
+                ->setParameter('{{ actualClass }}', get_class($value))
+                ->addViolation()
+            ;
+        }
+
+        $actualContentType = $value->getContentTypeName();
+        $expectedContentType = $this->context->getObject()->getContentTypeName();
+
+        if ($expectedContentType !== $actualContentType) {
```

Comment:
`expectedContentType` könnte auch `null` sein. In diesem Fall wird der contentType vom prototype kopiert.

Dann müsste man nur prüfen, dass `$prototype->contentType->entityClass` kompatibel ist mit `$object::class`. Das ist aber eigentlich durch den Test oben bereits abgedeckt.

- `$prototype->contentType->entityClass` ist kompatibel mit `$prototype::class` (kann gar nicht anders in der DB stehen)
- $prototype::class ist gleich wie $object::class (wird oben geprüft)

Reply:
Stimmt. Ich könnte höchstens die AssertContentTypeCompatible Validierung auf einen neuen Getter getContentTypeFromThisOrPrototype verschieben, dann wäre sogar inkonsistenter Datenbankzustand abgedeckt. Fändest du das sinnvoll?

Reply:
Finde ich einen guten Vorschlag. Und deckt den Edge-case oben auch ab.

Reply:
Ich habe das so ergänzt. Allerdings ist es im Moment doch noch nicht möglich, dass `expectedContentType` `null` ist, weil die Symfony AutoMapping Validierung auf `contentType` noch greift (weil die DB-Spalte non-nullable ist). Ich habe ein `DisableAutoMapping` auf `ContentNode::$contentType` hinzugefügt, aber wegen einem [Bug in Symfony](https://github.com/symfony/symfony/issues/35399) reicht das nicht, wir müssten das auf allen Subklassen auch nochmals schreiben. Stattdessen habe ich jetzt geskippte Tests dafür vorbereitet, die wir dann aktivieren können wenn Symfony den Bug gefixt hat. Ein [PR](https://github.com/symfony/symfony/pull/40294) dafür ist seit Februar offen, aber es ist anscheinend nicht ganz einfach, das zu fixen.